benchmark/benchmain/main.go

```diff
@@ -648,76 +659,39 @@ func setCompressorMode(val string) []string {
 	}
 }
 
-func printThroughput(requestCount uint64, requestSize int, responseCount uint64, responseSize int, benchTime time.Duration) {
-	requestThroughput := float64(requestCount) * float64(requestSize) * 8 / benchTime.Seconds()
-	responseThroughput := float64(responseCount) * float64(responseSize) * 8 / benchTime.Seconds()
-	fmt.Printf("Number of requests:  %v\tRequest throughput:  %v bit/s\n", requestCount, requestThroughput)
-	fmt.Printf("Number of responses: %v\tResponse throughput: %v bit/s\n", responseCount, responseThroughput)
-	fmt.Println()
-}
-
 func main() {
 	opts := processFlags()
 	before(opts)
-	s := stats.NewStats(numStatsBuckets)
-	s.SortLatency()
-	var memStats runtime.MemStats
-	var results testing.BenchmarkResult
-	var startAllocs, startBytes uint64
-	var startTime time.Time
-	var startTimer = func() {
-		runtime.ReadMemStats(&memStats)
-		startAllocs = memStats.Mallocs
-		startBytes = memStats.TotalAlloc
-		startTime = time.Now()
-	}
-	var stopTimer = func(count uint64) {
-		runtime.ReadMemStats(&memStats)
-		results = testing.BenchmarkResult{
-			N:         int(count),
-			T:         time.Since(startTime),
-			Bytes:     0,
-			MemAllocs: memStats.Mallocs - startAllocs,
-			MemBytes:  memStats.TotalAlloc - startBytes,
-		}
-	}
 
-	// Run benchmarks
-	resultSlice := []stats.BenchResults{}
+	s := stats.NewStats(numStatsBuckets)
 	featuresNum := makeFeaturesNum(opts)
-	sharedPos := sharedFeatures(featuresNum)
-	for _, benchFeature := range opts.generateFeatures(featuresNum) {
-		grpc.EnableTracing = benchFeature.EnableTrace
-		if benchFeature.EnableChannelz {
+	sf := sharedFeatures(featuresNum)
+	mode := ""
+	for _, bf := range opts.generateFeatures(featuresNum) {
+		var (
+			start  = func() { s.StartRun(mode, bf, sf) }
```

Comment:
I think start shouldn't be a closure. Because start can be called in another go routine concurrently and when it is executed mode and bf may already be changed in this thread.

Reply:
Each of the calls to `unaryBenchmark`, `streamBenchmark` and `unconstrainedStreamBenchmark` return only after they are completely done. So, I don't think there will be problem of `bf` changing in this thread. Although the comment makes sense, so I have modified the `start` closure to actually be bound with its arguments, so we don't have any confusion.

Thanks.